src/storage/storage.h

```diff
@@ -358,8 +365,43 @@ class Storage {
 
   rocksdb::WriteOptions default_write_opts_ = rocksdb::WriteOptions();
 
-  rocksdb::Status writeToDB(const rocksdb::WriteOptions &options, rocksdb::WriteBatch *updates);
+  rocksdb::Status writeToDB(engine::Context &ctx, const rocksdb::WriteOptions &options, rocksdb::WriteBatch *updates);
   void recordKeyspaceStat(const rocksdb::ColumnFamilyHandle *column_family, const rocksdb::Status &s);
 };
 
+struct Context {
+  engine::Storage *storage = nullptr;
+  const rocksdb::Snapshot *snapshot = nullptr;
+  std::unique_ptr<rocksdb::WriteBatchWithIndex> batch = nullptr;
+
+  rocksdb::ReadOptions GetReadOptions();
+  const rocksdb::Snapshot *GetSnapShot();
+  void SetLatestSnapshot();
+
+  explicit Context(engine::Storage *storage) : storage(storage), snapshot(storage->GetDB()->GetSnapshot()) {}
```

Comment:
Will the snapshot be used in every case? If not does it increase the cost of db operations?

Reply:
I don’t quite understand what “changing it to lazily get the context” means and why we need to do this.

Reply:
Getting a snapshot, would acquire a global "pinned" version in RocksDB, and affect data reclaiming. Maybe we can delaying the time to get the snapshot when first acquire a read 🤔